# Import data with Pandas
As you can see from various examples, it is so tedious to start creating DataFrame from scratch. In the data science world, data will be available for you on a spreadsheet such as `MS-Excel`. Our job as a data scientist is to import those datasets using `Pandas`. Pandas can import different format of data such as `.xlsx`(Excel data), `.csv`(comma separated value), `.sav`(SPSS data), `.dat`(STATA data), or any other data online (the web.)

## How to read in data

It’s quite simple to load data from various file formats into a DataFrame. In this section, you will learn how to import different file format with `Pandas`.

![](images/Import_file.jpg)

# Import data from CSV


With `.csv` (comma-separated value) format all you need is `pd.read_csv()` command to load in the data as a DataFrame.

## Example 1

Load the Ethiopian's flowers dataset (`Ethiopian_flowers.csv`) that was adapted from the well known `Iris dataset`. Remember the first thing is to import all the necessary libraries such as numpy, pandas, etc. 

**Note**: The data is in the datasets folder or directory.

![](images/Ethiopian_flower.jpg)

In [1]:
import numpy as np
import pandas as pd

In [2]:
ethiopian_flower = pd.read_csv("datasets/Ethiopian_flowers.csv")

In [3]:
ethiopian_flower.head()

,Sepal.Length,Sepal.Width,Petal.Length,Petal.Width,Species
0,6.1,2.8,4.7,1.2,Carnation
1,5.9,3.0,5.1,1.8,Lily
2,6.0,3.4,4.5,1.6,Carnation
3,5.7,2.5,5.0,2.0,Lily
4,6.9,3.1,5.4,2.1,Lily


You have imported Ethiopian flowers successfully! Bravo! Henceforth, we you import a dataset, the result is always a `DataFrame`.

In [4]:
type(ethiopian_flower)

pandas.core.frame.DataFrame

So therefore, all the pandas attributes that have been discussed so far can be applied.

Let's see what is in ethiopian_flower that we just imported. Functions such as `.head()`, `.tail()`, `.shape`, etc. are useful in exploring our dataset. 

In [5]:
ethiopian_flower.head(10) # To see the first ten observations

,Sepal.Length,Sepal.Width,Petal.Length,Petal.Width,Species
0,6.1,2.8,4.7,1.2,Carnation
1,5.9,3.0,5.1,1.8,Lily
2,6.0,3.4,4.5,1.6,Carnation
3,5.7,2.5,5.0,2.0,Lily
4,6.9,3.1,5.4,2.1,Lily
5,5.7,2.8,4.1,1.3,Carnation
6,6.1,3.0,4.9,1.8,Lily
7,6.2,2.2,4.5,1.5,Carnation
8,6.7,2.5,5.8,1.8,Lily
9,5.7,2.6,3.5,1.0,Carnation


In [7]:
ethiopian_flower.tail() # To see the last five observations

,Sepal.Length,Sepal.Width,Petal.Length,Petal.Width,Species
345,7.2,3.0,5.8,1.6,Lily
346,6.7,3.1,4.7,1.5,Carnation
347,7.0,3.2,4.7,1.4,Carnation
348,5.6,3.0,4.5,1.5,Carnation
349,4.8,3.0,1.4,0.3,Rose


In [8]:
ethiopian_flower.shape

(350, 5)

Ethiopian flower dataset has 350 observations/rows with 5 variables/columns

In [9]:
ethiopian_flower.columns # list the names of the columns in the dataset

Index(['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width',
       'Species'],
      dtype='object')

The variables in the dataset are Sepal.Length, Sepal.Width, Petal.Length, Petal.Width, and Species.

In [10]:
ethiopian_flower.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Sepal.Length  350 non-null    float64
 1   Sepal.Width   350 non-null    float64
 2   Petal.Length  350 non-null    float64
 3   Petal.Width   350 non-null    float64
 4   Species       350 non-null    object 
dtypes: float64(4), object(1)
memory usage: 13.8+ KB


`.info()` attribute shows us that `ethiopian_flower` dataset has $4$ variables are float datatypes and $1$ variable that is object (or string) datatype.

## Importing data from Excel


With `.xlsx` or `xls` format, all you need is `pd.read_excel()` command to load in the data

## Example 2

Load the `East-Africa-names.xlsx` which was scraped from <https://www.behindthename.com/submit/names/usage/eastern-african>. Since you have imported all the necessary libraries such as numpy and pandas, you don't need to import them here again.

![](images/East_Africa_names.jpg)

Excel Workbook can have many worksheets and each can host different data. While importing your data, you need to specify the index of the sheet that you want to import. Remember Python index starts from $0$.

In [13]:
east_africa_names = pd.read_excel("datasets/East-Africa-names.xlsx", sheet_name= 0)

You have imported East Africa names successfully! You are now an expert in importing file!

Let's see various names in the east_africa_names data that we just imported. 

In [14]:
east_africa_names.head()

,Name,Gender,Origin.1,Origin.2,Origin.3,Origin.4
0,Aadaan,M,Somali,NaN,NaN,NaN
1,Aadam,M,Urdu,Somali,Estonian,NaN
2,Aaden,M,Somali,NaN,NaN,NaN
3,Aamiina,F,Somali,NaN,NaN,NaN
4,Abadir,M,Arabic,Bohairic,Sahidic,NaN


As you can see, a given name can come from more than 1 origin and also both sex can bear that name. The `NaN` in some of the columns means that the values are not available.

## Import data from JSON 

JSON stands for JavaScript Object Notation. With `.json`format, all you need is `pd.read_json()` command to load in the data

## Example 4

Load the `Ugandan_women.json` data. This data set gives the average heights and weights for Ugandan women aged $30$–$39$.

In [15]:
ugandan_women = pd.read_json("datasets/Ugandan_women.json")

In [16]:
ugandan_women.head()

,height,weight
0,63,129
1,69,150
2,71,159
3,68,146
4,58,115


In [17]:
ugandan_women.tail()

,height,weight
295,65,135
296,62,126
297,66,139
298,58,115
299,58,115


In [18]:
ugandan_women.shape

(300, 2)

The dataset consist of 300 rows and 2 columns

# Export data with Pandas

You can save DataFrame in Python to any file of your choice. Similar to the ways we read in data, pandas provides intuitive commands to save it:

`df.to_csv('new_file.csv')` # to save as csv file

`df.to_excel('new_file.xlsx')` # to save as Excel file

`df.to_json('new_file.json')` # to save as JSON file

When we save as .`csv`, `.xlsx`, or any file formats, all we have to input into those functions is our desired filename with the appropriate file extension.

# Example 

Import `fatal-police-shootings-data.csv` data as `police_shooting` and export the resulting DataFrame as:
    
1. `.dta`

2. `.xlsx`

3. `.json`

# Solution

In [19]:
import pandas as pd

In [20]:
police_shooting = pd.read_csv("datasets/fatal-police-shootings-data.csv")

In [21]:
police_shooting.head() # To see the first 5 observation

,date,manner_of_death,armed,age,gender,race,city,state,signs_of_mental_illness,threat_level,flee,body_camera,longitude,latitude
0,02-01-15,shot,gun,53.0,M,A,Shelton,WA,True,attack,Not fleeing,False,-123.122,47.247
1,02-01-15,shot,gun,47.0,M,W,Aloha,OR,False,attack,Not fleeing,False,-122.892,45.487
2,03-01-15,shot and Tasered,unarmed,23.0,M,H,Wichita,KS,False,other,Not fleeing,False,-97.281,37.695
3,04-01-15,shot,toy weapon,32.0,M,W,San Francisco,CA,True,attack,Not fleeing,False,-122.422,37.763
4,04-01-15,shot,nail gun,39.0,M,H,Evans,CO,False,attack,Not fleeing,False,-104.692,40.384


In [22]:
police_shooting.shape # To check the dimension of the dataset

(5793, 14)

In [23]:
police_shooting.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5793 entries, 0 to 5792
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   date                     5793 non-null   object 
 1   manner_of_death          5793 non-null   object 
 2   armed                    5580 non-null   object 
 3   age                      5547 non-null   float64
 4   gender                   5792 non-null   object 
 5   race                     5256 non-null   object 
 6   city                     5793 non-null   object 
 7   state                    5793 non-null   object 
 8   signs_of_mental_illness  5793 non-null   bool   
 9   threat_level             5793 non-null   object 
 10  flee                     5486 non-null   object 
 11  body_camera              5793 non-null   bool   
 12  longitude                5510 non-null   float64
 13  latitude                 5510 non-null   float64
dtypes: bool(2), float64(3), 

**To export as `dta` (STATA file), we use:**

In [24]:
police_shooting.to_stata("datasets/police_shooting.dta")

**To export as `.xlsx` (Excel file), we use:**

In [25]:
police_shooting.to_excel("datasets/police_shooting.xlsx")

**To export as `.json` (JSON file), we use:**

In [26]:
police_shooting.to_json("datasets/police_shooting.json")

We then check our working directory for all the data that we just exported.

![](images/working_dir3.png)

In [29]:
police_shooting.to_csv("datasets/PoliceShooting.csv")